File: src/papyrus_api/routers/providers.py
"""Provider management API routes.

This module provides CRUD operations for AI providers, API keys, and models.
All API keys are encrypted at rest.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from papyrus.paths import DATABASE_FILE
from papyrus.data.database import (
    delete_api_key,
    delete_model,
    delete_provider,
    load_all_providers,
    save_api_key,
    save_model,
    save_provider,
    set_default_provider,
    update_provider_enabled,
)
from papyrus.data.crypto import decrypt_api_key, encrypt_api_key

router = APIRouter(tags=["providers"])


# ============================================================================
# Pydantic Models
# ============================================================================

class ApiKeyItem(BaseModel):
    id: str = ""
    name: str = "default"
    key: str = ""  # Plain text key, will be encrypted before storage


class ProviderModel(BaseModel):
    id: str = ""
    name: str = ""
    modelId: str = ""
    port: str = "openai"
    capabilities: list[str] = []
    apiKeyId: str | None = None
    enabled: bool = True


class Provider(BaseModel):
    id: str = ""
    type: str = "custom"
    name: str = ""
    baseUrl: str = ""
    enabled: bool = False
    isDefault: bool = False
    apiKeys: list[ApiKeyItem] = []
    models: list[ProviderModel] = []


class ProviderListResponse(BaseModel):
    success: bool
    providers: list[Provider]


class ProviderResponse(BaseModel):
    success: bool
    provider: Provider | None = None
    message: str = ""


class ProviderUpdateRequest(BaseModel):
    name: str | None = None
    baseUrl: str | None = None
    enabled: bool | None = None
    apiKeys: list[ApiKeyItem] | None = None


class ModelUpdateRequest(BaseModel):
    name: str
    modelId: str
    port: str = "openai"
    capabilities: list[str] = []
    apiKeyId: str | None = None
    enabled: bool = True


# ============================================================================
# Provider CRUD Endpoints
# ============================================================================

@router.get("/providers", response_model=ProviderListResponse)
def get_providers() -> ProviderListResponse:
    """Get all providers with their API keys and models."""
    try:
        providers_data = load_all_providers(DATABASE_FILE)
        
        providers = []
        for p in providers_data:
            providers.append(Provider(
                id=p["id"],
                type=p["type"],
                name=p["name"],
                baseUrl=p["baseUrl"],
                enabled=p["enabled"],
                isDefault=p["isDefault"],
                apiKeys=[ApiKeyItem(**k) for k in p["apiKeys"]],
                models=[ProviderModel(**m) for m in p["models"]],
            ))
        
        return ProviderListResponse(success=True, providers=providers)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to load providers: {e}")


@router.post("/providers", response_model=ProviderResponse)
def create_provider(provider: Provider) -> ProviderResponse:
    """Create a new provider with API keys and models."""
    try:
        # Save provider
        provider_id = save_provider(DATABASE_FILE, {
            "id": provider.id or None,
            "type": provider.type,
            "name": provider.name,
            "baseUrl": provider.baseUrl,
            "enabled": provider.enabled,
            "isDefault": provider.isDefault,
        })
        
        # Save API keys
        for key in provider.apiKeys:
            save_api_key(DATABASE_FILE, provider_id, {
                "id": key.id or None,
                "name": key.name,
                "key": key.key,
            })
        
        # Save models
        for model in provider.models:
            save_model(DATABASE_FILE, provider_id, {
                "id": model.id or None,
                "name": model.name,
                "modelId": model.modelId,
                "port": model.port,
                "capabilities": model.capabilities,
                "apiKeyId": model.apiKeyId,
                "enabled": model.enabled,
            })
        
        # Return updated provider
        providers = load_all_providers(DATABASE_FILE)
        new_provider = next((p for p in providers if p["id"] == provider_id), None)
        
        if new_provider:
            return ProviderResponse(
                success=True,
                provider=Provider(**new_provider),
                message="Provider created successfully"
            )
        else:
            return ProviderResponse(success=False, message="Provider created but not found")
    
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create provider: {e}")


@router.put("/providers/{provider_id}", response_model=ProviderResponse)
def update_provider(provider_id: str, updates: ProviderUpdateRequest) -> ProviderResponse:
    """Update a provider's basic info and API keys."""
    try:
        # Update provider basic info
        provider_data = {
            "id": provider_id,
        }
        if updates.name is not None:
            provider_data["name"] = updates.name
        if updates.baseUrl is not None:
            provider_data["baseUrl"] = updates.baseUrl
        if updates.enabled is not None:
            provider_data["enabled"] = updates.enabled
        
        save_provider(DATABASE_FILE, provider_data)
        
        # Update API keys if provided
        if updates.apiKeys is not None:
            # Get existing keys
            providers = load_all_providers(DATABASE_FILE)
            existing_provider = next((p for p in providers if p["id"] == provider_id), None)
            existing_key_ids = {k["id"] for k in (existing_provider.get("apiKeys", []) if existing_provider else [])}
            new_key_ids = {k.id for k in updates.apiKeys if k.id}
            
            # Delete keys that are no longer present
            for key_id in existing_key_ids - new_key_ids:
                delete_api_key(DATABASE_FILE, key_id)
            
            # Save/update keys
            for key in updates.apiKeys:
                save_api_key(DATABASE_FILE, provider_id, {
                    "id": key.id or None,
                    "name": key.name,
                    "key": key.key,
                })
        
        # Return updated provider
        providers = load_all_providers(DATABASE_FILE)
        updated = next((p for p in providers if p["id"] == provider_id), None)
        
        if updated:
            return ProviderResponse(
                success=True,
                provider=Provider(**updated),
                message="Provider updated successfully"
            )
        else:
            return ProviderResponse(success=False, message="Provider not found")
    
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update provider: {e}")


@router.delete("/providers/{provider_id}")
def delete_provider_endpoint(provider_id: str) -> dict[str, Any]:
    """Delete a provider and all its data."""
    try:
        success = delete_provider(DATABASE_FILE, provider_id)
        if success:
            return {"success": True, "message": "Provider deleted successfully"}
        else:
            return {"success": False, "message": "Provider not found"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete provider: {e}")


@router.post("/providers/{provider_id}/default")
def set_provider_default(provider_id: str) -> dict[str, Any]:
    """Set a provider as the default."""
    try:
        success = set_default_provider(DATABASE_FILE, provider_id)
        if success:
            return {"success": True, "message": "Default provider set successfully"}
        else:
            return {"success": False, "message": "Provider not found"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to set default provider: {e}")


@router.post("/providers/{provider_id}/enabled")
def set_provider_enabled(provider_id: str, enabled: dict[str, bool]) -> dict[str, Any]:
    """Enable or disable a provider."""
    try:
        success = update_provider_enabled(DATABASE_FILE, provider_id, enabled.get("enabled", False))
        if success:
            return {"success": True, "message": "Provider status updated"}
        else:
            return {"success": False, "message": "Provider not found"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update provider status: {e}")


# ============================================================================
# Model Management Endpoints
# ============================================================================

@router.post("/providers/{provider_id}/models", response_model=ProviderResponse)
def add_model(provider_id: str, model: ModelUpdateRequest) -> ProviderResponse:
    """Add a model to a provider."""
    try:
        save_model(DATABASE_FILE, provider_id, {
            "name": model.name,
            "modelId": model.modelId,
            "port": model.port,
            "capabilities": model.capabilities,
            "apiKeyId": model.apiKeyId,
            "enabled": model.enabled,
        })
        
        # Return updated provider
        providers = load_all_providers(DATABASE_FILE)
        updated = next((p for p in providers if p["id"] == provider_id), None)
        
        if updated:
            return ProviderResponse(
                success=True,
                provider=Provider(**updated),
                message="Model added successfully"
            )
        else:
            return ProviderResponse(success=False, message="Provider not found")
    
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to add model: {e}")


@router.put("/providers/{provider_id}/models/{model_id}", response_model=ProviderResponse)
def update_model(provider_id: str, model_id: str, model: ModelUpdateRequest) -> ProviderResponse:
    """Update a provider model."""
    try:
        save_model(DATABASE_FILE, provider_id, {
            "id": model_id,
            "name": model.name,
            "modelId": model.modelId,
            "port": model.port,
            "capabilities": model.capabilities,
            "apiKeyId": model.apiKeyId,
            "enabled": model.enabled,
        })
        
        # Return updated provider
        providers = load_all_providers(DATABASE_FILE)
        updated = next((p for p in providers if p["id"] == provider_id), None)
        
        if updated:
            return ProviderResponse(
                success=True,
                provider=Provider(**updated),
                message="Model updated successfully"
            )
        else:
            return ProviderResponse(success=False, message="Provider not found")
    
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update model: {e}")


@router.delete("/providers/{provider_id}/models/{model_id}")
def delete_model_endpoint(provider_id: str, model_id: str) -> dict[str, Any]:
    """Delete a model from a provider."""
    try:
        success = delete_model(DATABASE_FILE, model_id)
        if success:
            return {"success": True, "message": "Model deleted successfully"}
        else:
            return {"success": False, "message": "Model not found"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete model: {e}")


# ============================================================================
# API Key Management Endpoints (Individual)
# ============================================================================

@router.post("/providers/{provider_id}/apikeys")
def add_api_key(provider_id: str, key: ApiKeyItem) -> dict[str, Any]:
    """Add an API key to a provider."""
    try:
        key_id = save_api_key(DATABASE_FILE, provider_id, {
            "name": key.name,
            "key": key.key,
        })
        return {"success": True, "keyId": key_id, "message": "API key added"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to add API key: {e}")


@router.delete("/providers/{provider_id}/apikeys/{key_id}")
def delete_api_key_endpoint(provider_id: str, key_id: str) -> dict[str, Any]:
    """Delete an API key."""
    try:
        success = delete_api_key(DATABASE_FILE, key_id)
        if success:
            return {"success": True, "message": "API key deleted"}
        else:
            return {"success": False, "message": "API key not found"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete API key: {e}")


# ============================================================================
# Utility Endpoints
# ============================================================================

# SECURITY: test-decrypt endpoint removed to prevent decryption oracle attacks.
# Encrypted API keys in the database should never be decryptable via API.

File: backend/src/api/routes/providers.ts
import type { FastifyInstance } from 'fastify';
import { z } from 'zod';
import {
  loadAllProviders,
  saveProvider,
  deleteProvider,
  setDefaultProvider,
  updateProviderEnabled,
  saveApiKey,
  deleteApiKey,
  saveModel,
  deleteModel,
  runInTransaction,
} from '../../db/database.js';
import { getDb } from '../../db/database.js';
import type { Provider } from '../../core/types.js';
import { aiConfig } from '../../ai/config-instance.js';
import { loadAIConfigFromDb } from '../../ai/db-sync.js';

const ApiKeySchema = z.object({
  id: z.string().optional(),
  name: z.string().min(1),
  key: z.string(),
});

const ModelSchema = z.object({
  id: z.string().optional(),
  name: z.string().min(1),
  modelId: z.string().min(1),
  port: z.string().optional(),
  capabilities: z.array(z.string()).optional(),
  apiKeyId: z.string().nullable().optional(),
  enabled: z.boolean().optional(),
});

const ProviderSchema = z.object({
  id: z.string().optional(),
  type: z.string().optional(),
  name: z.string().min(1),
  baseUrl: z.string().optional(),
  enabled: z.boolean().optional(),
  isDefault: z.boolean().optional(),
  apiKeys: z.array(ApiKeySchema).optional(),
  models: z.array(ModelSchema).optional(),
}).passthrough();

export default async function providersRoutes(fastify: FastifyInstance): Promise<void> {
  fastify.get('/', async (request, reply) => {
    try {
      const providers = loadAllProviders();
      reply.send({ success: true, providers });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.post('/', async (request, reply) => {
    const parseResult = ProviderSchema.safeParse(request.body);
    if (!parseResult.success) {
      reply.status(400).send({ success: false, error: parseResult.error.errors.map(e => e.message).join('; ') });
      return;
    }
    const body = parseResult.data as Partial<Provider>;
    try {
      const id = runInTransaction(() => {
        const providerId = saveProvider(body);

        if (body.apiKeys) {
          for (const key of body.apiKeys) {
            saveApiKey(providerId, key);
          }
        }
        if (body.models) {
          for (const model of body.models) {
            saveModel(providerId, model);
          }
        }

        return providerId;
      });

      reply.send({ success: true, provider: { ...body, id }, message: 'Provider created' });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.includes('UNIQUE constraint failed')) {
        reply.status(409).send({ success: false, error: '相同配置的服务商已存在' });
        return;
      }
      reply.status(500).send({ success: false, error: `添加供应商失败: ${msg}` });
    }
  });

  fastify.put('/:providerId', async (request, reply) => {
    const { providerId } = request.params as { providerId: string };
    const parseResult = ProviderSchema.safeParse(request.body);
    if (!parseResult.success) {
      reply.status(400).send({ success: false, error: parseResult.error.errors.map(e => e.message).join('; ') });
      return;
    }
    const body = parseResult.data as Partial<Provider>;
    try {
      runInTransaction(() => {
        saveProvider({ ...body, id: providerId });

        const incomingKeyIds = new Set(body.apiKeys?.map(k => k.id).filter(Boolean) ?? []);
        if (body.apiKeys) {
          for (const key of body.apiKeys) {
            const savedKeyId = saveApiKey(providerId, key);
            incomingKeyIds.add(savedKeyId);
          }
        }
        const db = getDb();
        const allKeys = db.prepare('SELECT id FROM api_keys WHERE provider_id = ?').all(providerId) as { id: string }[];
        for (const row of allKeys) {
          if (!incomingKeyIds.has(row.id)) {
            deleteApiKey(row.id);
          }
        }
      });
      reply.send({ success: true, message: 'Provider updated' });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.includes('UNIQUE constraint failed')) {
        reply.status(409).send({ success: false, error: '相同配置的服务商已存在' });
        return;
      }
      reply.status(500).send({ success: false, error: `更新供应商失败: ${msg}` });
    }
  });

  fastify.delete('/:providerId', async (request, reply) => {
    try {
      const { providerId } = request.params as { providerId: string };
      const success = deleteProvider(providerId);
      if (!success) {
        reply.status(404).send({ success: false, error: 'Provider not found' });
        return;
      }
      reply.send({ success: true, message: 'Provider deleted' });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.post('/:providerId/default', async (request, reply) => {
    try {
      const { providerId } = request.params as { providerId: string };
      setDefaultProvider(providerId);
      // 从数据库加载最新配置到内存
      loadAIConfigFromDb(aiConfig, true);
      reply.send({ success: true, message: 'Default provider set' });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  const EnabledSchema = z.object({ enabled: z.boolean() });

  fastify.post('/:providerId/enabled', async (request, reply) => {
    try {
      const { providerId } = request.params as { providerId: string };
      const parseResult = EnabledSchema.safeParse(request.body);
      if (!parseResult.success) {
        reply.status(400).send({ success: false, error: parseResult.error.errors.map(e => e.message).join('; ') });
        return;
      }
      updateProviderEnabled(providerId, parseResult.data.enabled);
      reply.send({ success: true, message: 'Provider enabled status updated' });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.post('/:providerId/models', async (request, reply) => {
    const { providerId } = request.params as { providerId: string };
    const parseResult = ModelSchema.safeParse(request.body);
    if (!parseResult.success) {
      reply.status(400).send({ success: false, error: parseResult.error.errors.map(e => e.message).join('; ') });
      return;
    }
    const body = parseResult.data;
    try {
      const modelId = saveModel(providerId, body);
      reply.send({ success: true, modelId, message: 'Model added' });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.includes('UNIQUE constraint failed')) {
        reply.status(409).send({ success: false, error: '该模型已存在于当前供应商' });
        return;
      }
      if (msg.includes('FOREIGN KEY')) {
        reply.status(400).send({ success: false, error: `添加模型失败:外键约束失败,apiKeyId 或 providerId 不存在 (${msg})` });
        return;
      }
      reply.status(500).send({ success: false, error: `添加模型失败: ${msg}` });
    }
  });

  fastify.put('/:providerId/models/:modelId', async (request, reply) => {
    const { providerId, modelId } = request.params as { providerId: string; modelId: string };
    const parseResult = ModelSchema.safeParse(request.body);
    if (!parseResult.success) {
      reply.status(400).send({ success: false, error: parseResult.error.errors.map(e => e.message).join('; ') });
      return;
    }
    const body = parseResult.data;
    try {
      saveModel(providerId, { ...body, id: modelId });
      reply.send({ success: true, message: 'Model updated' });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.includes('FOREIGN KEY')) {
        reply.status(400).send({ success: false, error: `更新模型失败:外键约束失败,apiKeyId 或 providerId 不存在 (${msg})` });
        return;
      }
      reply.status(500).send({ success: false, error: `更新模型失败: ${msg}` });
    }
  });

  fastify.delete('/:providerId/models/:modelId', async (request, reply) => {
    try {
      const { modelId } = request.params as { modelId: string };
      const deleted = deleteModel(modelId);
      if (!deleted) {
        reply.status(404).send({ success: false, error: 'Model not found' });
        return;
      }
      reply.send({ success: true, message: 'Model deleted' });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.post('/:providerId/apikeys', async (request, reply) => {
    try {
      const { providerId } = request.params as { providerId: string };
      const parseResult = ApiKeySchema.safeParse(request.body);
      if (!parseResult.success) {
        reply.status(400).send({ success: false, error: parseResult.error.errors.map(e => e.message).join('; ') });
        return;
      }
      const body = parseResult.data;
      const keyId = saveApiKey(providerId, body);
      reply.send({ success: true, keyId, message: 'API key added' });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });

  fastify.delete('/:providerId/apikeys/:keyId', async (request, reply) => {
    try {
      const { keyId } = request.params as { keyId: string };
      deleteApiKey(keyId);
      reply.send({ success: true, message: 'API key deleted' });
    } catch (err) {
      const message = err instanceof Error ? err.message : '服务器内部错误';
      request.log.error({ err }, message);
      reply.status(500).send({ success: false, error: message });
    }
  });
}
